src/frontend/src/expr/type_inference/func.rs

```diff
@@ -580,79 +580,54 @@ fn infer_type_for_special(
         }
         ExprType::ArrayDistinct => {
             ensure_arity!("array_distinct", | inputs | == 1);
-            let ret_type = inputs[0].return_type();
-            if inputs[0].is_untyped() {
-                return Err(ErrorCode::BindError(
-                    "could not determine polymorphic type because input has type unknown"
-                        .to_string(),
-                )
-                .into());
-            }
-            match ret_type {
-                DataType::List(list_elem_type) => Ok(Some(DataType::List(list_elem_type))),
-                _ => Ok(None),
-            }
+            inputs[0].ensure_array_type()?;
+
+            Ok(Some(inputs[0].return_type()))
         }
         ExprType::ArrayDims => {
             ensure_arity!("array_dims", | inputs | == 1);
```

Comment:
nit: maybe we can also change `ensure_arity` to a function accepting a range? The macro seems unnecessary.

Reply:
Makes sense 😂 I will do it when appropriate.